### ISSUE TITLE
[wiki] Building on Ubuntu 19.04 x86_64 (not a bug, a tutorial i thought i'd share)

### ISSUE THREAD

**Issue by rocket-pig** [quiet/quiet#48]
I spent a while getting here and these days I meticulously document everything so I don't have to re-figure it out two weeks later.  I thought I'd share, maybe you'll put it on your wiki page or somewhere.  Thank you so much for sharing 'quiet' with simpler folks like me <3

# Building everything including deps from source
having the bajillion standard tools required for a common build environment not included here.

## building deps:
    mkdir quiet-project
    mkdir required-deps
    cd into required-deps and read on:

### libsndfile: (if you want to encode/decode to/from .wav files)
        wget "http://www.mega-nerd.com/libsndfile/files/libsndfile-1.0.28.tar.gz"
        tar xzvf libsndfile-1.0.28.tar.gz
        cd libsndfile-1.0.28/
        ./configure;make;sudo make install

### liquid-dsp:
        unzip liquid-dsp-devel.zip
        (OR  git clone https://github.com/quiet/liquid-dsp.git -b devel --single-branch)
        cd liquid-dsp
        ./bootstrap
        ./configure CFLAGS="-O -msse3" #youre fucked without the cflags, its a known bug since 2016 amazingly
        make
        sudo make install

        it only installs three things:
        libliquid.so libliquid.a in /usr/local/lib
        liquid.h in /usr/local/include/liquid

### jansson:
        mkdir jansson; cd jansson
        wget "https://github.com/akheron/jansson/archive/v2.12.tar.gz"
        tar xzvf v2.12.tar.gz
        cd jans
        autoreconf -i
        ./configure;make;sudo make install
        sudo ldconfig

### building 'quiet' 
    tar xzvf quiet.tar.gz  #(or download with git clone, i saved it locally to prevent changes from screwing me up later when attempting cross-compile)
    cd quiet
    ./bootstrap
    cd build
    sudo make install
    output of that:
        Install the project...
        -- Install configuration: ""
        -- Installing: /usr/local/share/quiet/quiet-profiles.json
        -- Up-to-date: /usr/local/include
        -- Installing: /usr/local/include/quiet.h
        -- Installing: /usr/local/include/quiet-portaudio.h
        -- Installing: /usr/local/bin/quiet_encode_soundcard
        -- Set runtime path of "/usr/local/bin/quiet_encode_soundcard" to ""
        -- Installing: /usr/local/bin/quiet_decode_soundcard
        -- Set runtime path of "/usr/local/bin/quiet_decode_soundcard" to ""
        -- Installing: /usr/local/lib/libquiet.so
        -- Set runtime path of "/usr/local/lib/libquiet.so" to "@rpath/libquiet.dylib"
        -- Installing: /usr/local/lib/libquiet.a

*note* 'quiet_encode_soundcard' and 'quiet_decode_soundcard' were both added to /usr/local/bin/
    at this point, one tweaks (if they wish) individual 'profile' settings in 'quiet-profiles.json'
    and then calls encode/decode <profilename> [<input_source>]
    the available profiles list:
        audible-7k-channel-0
        audible-7k-channel-1
        hello-world
        ultrasonic
        ultrasonic-3600
        cable-64k
        audible
        ultrasonic-whisper

#### final forward-error-correction caveat:
it _says_ that libfec is not required. but then when i read in the 
bugs section, the author says 'you basically need it' and that all profiles that define
either a 'inner_fec_scheme' or 'outer_fec_scheme' will crash if libfec isnt around.
after opening up /usr/local/share/quiet/quiet-profiles.json and changing the 'audible' profile to have both fec lines = 'none', one can successfully call 'quiet_encode_soundcard audible /path/to/textfile' and hear the encoding coming from speakers! woot!

### some example uses:
**play audible encoding directly from speakers**
    `quiet_encode_soundcard audible /tmp/textfile.txt`
**decode same (on different machine/soundcard) and print to file:**
    `quiet_decode_soundcard audible /tmp/decoded.txt`

**encode _inaudibly_ to file: (always saves in 'encoded.wav' in current dir)**
    `quiet_encode_file ultrasonic /tmp/textfile.txt`
**decode same:**
    `quiet_decode_file ultrasonic encoded.wav /tmp/decoded.txt`
____
**edit:** added some example uses.

**Comment by brian-armstrong** [comment]
This looks nice, thank you! I'll take a closer look later. One other thing to note is that now you can use libcorrect + the fec shim instead of libfec. Libcorrect should generally build a little more modernly. And also the travis.yml file for this repo might provide some help for how I've found to build things.

**Comment by rocket-pig** [comment]
Cool glad its appreciated, the thanks are all for you though! 

I hoped to teach myself cross compilation and get your 'quiet' going on an old iOS device.  It may be out of my league, though I have certainly poured a lot of energy into it. I'll be sure to share notes on that if I ever do. 

edit to clarify i meant cross-compilation /without/ Xcode. or a Mac (..or a developer login, etc).  I see someone has made a 'quiet' iOS app.  https://github.com/winster/QuietShare anywho that's my own tangent.  Thanks again and glad to be of any tiny help :)

**Comment by brian-armstrong** [comment]
Yep, I've made bindings for it. If you're curious, the iOS build script for the C portions is here https://github.com/quiet/QuietModemKit/blob/master/Build-Phases/build-c-libraries.sh

**Comment by rocket-pig** [comment]
Very cool, I had not previously found that.  Ive spent well beyond rational amounts of time trying to get cross compilation to work.  I see other projects release iOS binaries (aubio is one).  Will you ever consider doing so?
